src/table/use-sticky-columns-context.tsx

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0
+import React, { createContext, useContext, useLayoutEffect, useCallback, useRef, ReactNode } from 'react';
+import { unstable_batchedUpdates } from 'react-dom';
+
+interface StickyColumnsContextValue {
+  subscribe: (callback: (entry: any) => void) => void;
+  unsubscribe: (callback: (entry: any) => void) => void;
+}
+
+const StickyColumnsContext = createContext<StickyColumnsContextValue | null>(null);
+
+export const useStickyColumnsContext = () => {
+  const context = useContext(StickyColumnsContext);
+  if (!context) {
+    return { subscribe: () => {}, unsubscribe: () => {} };
+  }
+  const { subscribe, unsubscribe } = context as StickyColumnsContextValue;
+  return { subscribe, unsubscribe };
+};
+
+interface StickyColumnsContextProviderProps {
+  children: ReactNode;
+  wrapperRef: React.RefObject<HTMLDivElement>;
+  tableRef: React.RefObject<HTMLTableElement>;
+  shouldDisableStickyColumns: boolean;
+}
+interface StickyColumnsContextProvider {
+  (props: StickyColumnsContextProviderProps): JSX.Element;
+}
+
+export const StickyColumnsContextProvider: StickyColumnsContextProvider = ({
+  children,
+  wrapperRef,
+  tableRef,
+  shouldDisableStickyColumns,
+}) => {
+  const callbacks = useRef<Set<(entry: any) => void>>(new Set());
+
+  const checkStickyState = React.useCallback(() => {
+    const wrapper = wrapperRef.current;
+    const table = tableRef.current;
+    if (wrapper && table) {
+      const tableLeftPadding = Number(getComputedStyle(table).paddingLeft.slice(0, -2));
+      const tableRightPadding = Number(getComputedStyle(table).paddingRight.slice(0, -2));
+      const right = wrapper.scrollLeft < wrapper.scrollWidth - wrapper.clientWidth - tableRightPadding;
+      const left = wrapper.scrollLeft > tableLeftPadding;
+      return { left, right };
+    }
+    return { left: false, right: false };
+  }, [wrapperRef, tableRef]);
+
+  const createScrollHandler = useCallback(() => {
+    const scrollHandler = () => {
+      const { left, right } = checkStickyState();
+      // Call all subscribed callbacks
+      unstable_batchedUpdates(() => {
+        callbacks.current.forEach(callback => callback({ left, right }));
+      });
+    };
+    return scrollHandler;
+  }, [checkStickyState]);
+
+  useLayoutEffect(() => {
+    const scrollHandler = createScrollHandler();
+    const wrapper = wrapperRef.current;
+    const currentCallbacks = callbacks.current;
+    if (!shouldDisableStickyColumns) {
+      wrapper && wrapper.addEventListener('scroll', scrollHandler);
+      window.addEventListener('resize', scrollHandler);
+    } else if (shouldDisableStickyColumns) {
+      wrapper && wrapper.removeEventListener('scroll', scrollHandler);
+      window.removeEventListener('resize', scrollHandler);
+    }
+
+    return () => {
+      wrapper && wrapper.removeEventListener('scroll', scrollHandler);
+      window.removeEventListener('resize', scrollHandler);
+
+      currentCallbacks.clear();
```

Comment:
Would it be better to re-assign the callbacks ref? When we keep the reference there is a risk that not all callbacks will be cleared in case `subscribe` is called asynchronously.